### ISSUE TITLE
Build Modular Prediction Framework: Strategy/Registry Pattern + Markov Chain

### ISSUE THREAD

**Issue by cbwinslow** [cbwinslow/retrosheet#63]
## Context

Building on the established PA outcome model (HGB + advanced_count, log loss 1.5089), we now need a flexible, modular framework that:

1. Supports many granular targets (not just PA outcomes, but inning runs, strikeout rates, specific situational questions)
2. Pluggable model architectures (Markov chains, HGB, logistic regression, softmax regression)
3. Each prediction has clearly defined variables, no rigid schema

## Research Foundation

**Markov Chain Models (Established in Sabermetrics):**
- UT Austin (2016): "A Markov Chain Model for Predicting Major League Baseball" — treats baseball as discrete-time Markov chain with finite states
- Stanford CS229: Softmax regression for transition probabilities, beat Vegas over/under lines
- KCL Expected Runs: Markov chain expected runs matrix from real data
- mlb-win-probability repo: Ensemble combining RE24 + Markov + LightGBM + Bayesian (Brier 0.1605)

**Key Insight**: Markov property holds approximately — "future depends only on current state" — but context dependencies (pitcher fatigue, manager decisions) violate strict Markov.

## Proposed Architecture

**Model selection guidance:**
- PA outcome distribution → HGB, softmax (rich features, multi-class)
- Inning runs → Markov chain empirical (simple state-based)
- Next base-out state → Markov informed, softmax (discrete transitions)
- Strikeout rate → HGB + empirical (pitcher-specific features)

## Implementation Plan

1. **Create registry + base classes** (, )
2. **Port PA outcome** into new framework (validate pattern)
3. **Add run expectancy matrix** table in  (computed from Retrosheet)
4. **Add Markov chain model** as second model family
5. **Add  target** as second concrete target
6. **Document in KB**: Update KNOWLEDGE_BASE with Markov chain research

## Related Docs

- [docs/KNOWLEDGE_BASE_SABERMETRICS.md](docs/KNOWLEDGE_BASE_SABERMETRICS.md)
- [docs/KNOWLEDGE_BASE_MODELS_REPOS.md](docs/KNOWLEDGE_BASE_MODELS_REPOS.md)
- [docs/PREDICTION_ENGINE_PLAN.md](docs/PREDICTION_ENGINE_PLAN.md)
- [docs/agents/MODELING_WORKFLOWS.md](docs/agents/MODELING_WORKFLOWS.md)
- [docs/agents/CURRENT_SNAPSHOT.md](docs/agents/CURRENT_SNAPSHOT.md)

## Labels

enhancement, ml, predictions, research

**Comment by cbwinslow** [comment]
## Progress Update (2026-04-23)

### Completed:
Built modular prediction framework with Strategy pattern:

**Files created in `scripts/prediction_framework/`:**
- `base.py` - Predictor base class, PredictionTarget, ModelMetadata, PredictorRegistry
- `pa_predictor.py` - PAOutcomeDistributionPredictor implementation
- `__init__.py` - PredictionEngine unified interface
- `db.py` - Database config

**Database additions:**
- `models.register_model()` - register new models
- `models.register_calibration()` - register calibration artifacts  
- `models.promote_model()` - set active model
- `models.get_active_model()` - get active model for target
- `models.calibration_artifacts` table
- `models.v_active_models` view

**Active models registered:**
- `pa_outcome_distribution`: hist_gradient_boosting_multiclass (v20260412T045759Z)
- `game_home_win`: hist_gradient_boosting (v20260416T085327Z)
- `half_inning_any_run`: hist_gradient_boosting (v20260410T090317Z)
- `half_inning_lhb_any_hit`: hist_gradient_boosting (v20260410T090325Z)

### Usage:
```python
from prediction_framework import PredictionEngine

engine = PredictionEngine()
predictor = engine.get_predictor('pa_outcome_distribution')
result = predictor.predict(features_df)
```

### Next:
1. Add run expectancy matrix (#64)
2. Add Markov chain model
3. Add handedness matchup features
4. Research GIS pitch mapping

**Comment by cbwinslow** [comment]
## 🚀 Framework Code Committed (647a4fb)

The modular prediction framework has been scaffolded:

### New Files
- [scripts/prediction_framework/base.py](scripts/prediction_framework/base.py) — Base classes: Predictor, PredictionTarget, ModelMetadata, PredictorRegistry
- [scripts/prediction_framework/pa_predictor.py](scripts/prediction_framework/pa_predictor.py) — PAOutcomeDistributionPredictor implementation
- [scripts/prediction_framework/__init__.py](scripts/prediction_framework/__init__.py) — PredictionEngine unified interface
- [scripts/prediction_framework/db.py](scripts/prediction_framework/db.py) — Database configuration

### Active Models (as of 2026-04-23)
| Target | Model | Version |
|--------|-------|---------|
| pa_outcome_distribution | hist_gradient_boosting_multiclass | 20260412T045759Z |
| game_home_win | hist_gradient_boosting | 20260416T085327Z |
| half_inning_any_run | hist_gradient_boosting | 20260410T090317Z |
| half_inning_lhb_any_hit | hist_gradient_boosting | 20260410T090325Z |

### Next Steps
1. Port pa_outcome_distribution into new framework as validation
2. Add run expectancy matrix feature mart (#64)
3. Add Markov chain model as first new model family
4. Add next_state_distribution target as second concrete target

### Related Docs
- [docs/KNOWLEDGE_BASE_FRAMEWORK.md](docs/KNOWLEDGE_BASE_FRAMEWORK.md)
- [docs/KNOWLEDGE_BASE_MARKOV_CHAIN.md](docs/KNOWLEDGE_BASE_MARKOV_CHAIN.md)
- [docs/MODEL_SELECTION_GUIDE.md](docs/MODEL_SELECTION_GUIDE.md)

**Comment by cbwinslow** [comment]
Sub-issues created for modular prediction framework implementation:

- #74: Create predictions registry and base classes
- #76: Implement Markov Chain model family

These break the framework work into concrete deliverables. #74 is the first dependency — registry must exist before models can plug in.
